### ISSUE TITLE
BSD support

### ISSUE THREAD

**Issue by bkazemi** [scanmem/scanmem#179]

**Comment by bkazemi** [comment]
One issue I ran across is the use of `alloca()` for allocating variable length memory on the stack. This function adds some portability issues as GNU/Linux must include the non-standard header file `alloca.h`. Since `scanmem` is C99 code, I think just using variable length arrays instead would make portability a bit easier.

**Comment by bkazemi** [comment]
Scratch that last comment. `alloca()` is included in `stdlib.h` on FreeBSD. I say @sriemer 's solution is actually simpler.

**Comment by 12345ieee** [comment]
I'll chime in with a related question: is `scanmem` really C99?

Because the abundant use of `alloca()` screams `C89` to me, but on the other hand `void` pointer arithmetic is a GCC extension (https://gcc.gnu.org/onlinedocs/gcc-6.1.0/gcc/Pointer-Arith.html#Pointer-Arith), so maybe `gnu89`?

To confuse everything more, `gcc` now defaults to `gnu11`, so the Makefile without `-std=whatever` actually builds `scanmem` with the 2011 standard (unless your distro is stuck in the last decade).

@sriemer knows the answer for sure.

**Comment by bkazemi** [comment]
Well, I see some variadic macros here and there, so it is at least <something>99. Actually, I came across some of these macros in `scanmem.h` that use a custom name for `__VA_ARGS`, which I believe is a `gnu99` extension. Not quite sure on that one.

**Comment by sriemer** [comment]
Yes, scanmem is GNU C99 as it also uses GCC compiler specifics. I think this is just because of laziness. The legacy code is still quite a mess. There are e.g. inconsistent styles. Cleaning the mess up should be well planned as that breaks all patches some of us might have in the pipeline.

**Comment by bkazemi** [comment]
I'll open a new issue for the tracking of general code cleanup.

**Comment by bkazemi** [comment]
For anyone else who'd like to help with this, I've made a branch [here](https://github.com/bkazemi/scanmem/tree/bsd).

**Comment by bkazemi** [comment]
Update: picked this back up after some time away.

Gotten the bulk of this done, hit a snag with getting the virtual address of the stack and heap segments. Neither native FreeBSD `procfs` nor `linprocfs` give this information, AFAIK. @sriemer, maybe integrating `libelf` might prove beneficial if not for BSD only?

**Comment by bobbyhopere** [comment]
@bkazemi, on FreeBSD, the native way of getting the Virtual Memory Entries of a process in using **struct ptrace_vm_entry**. (It is not like in Linux where you read a file in /proc). See the **scan()** function in this example: https://github.com/troydhanson/uthash/blob/master/tests/hashscan.c#L477

Other related functions:
read_mem()
https://github.com/troydhanson/uthash/blob/master/tests/hashscan.c#L116

sigscan()
https://github.com/troydhanson/uthash/blob/master/tests/hashscan.c#L395

**Comment by bkazemi** [comment]
That makes things easy, thanks. I'll be sure to take a good look later.

**Comment by bkazemi** [comment]
I think there is probably too much to do here for `v0.17`. I've looked at the interface @bobbyhopere described, and you still can't see where the stack and heap regions are. If there is no way to find out, I will have to simply disallow searching in specific regions for BSD. Moving to `v0.18`.

**Comment by danfe** [comment]
Speaking of the `alloca.h`, it is actually checked by the autotools, and `HAVE_ALLOCA_H` macro is defined (or not) accordingly.  That said, consider the following patch:
```diff
--- common.h.orig
+++ common.h
@@ -28,7 +28,9 @@
 
 /* from string.h in glibc for Android/BSD */
 #ifndef strdupa
+# ifdef HAVE_ALLOCA_H
 # include <alloca.h>
+# endif
 # include <string.h>
 # define strdupa(s)                                                           \
     ({                                                                        \
@@ -40,7 +42,9 @@
 #endif
 
 #ifndef strndupa
+# ifdef HAVE_ALLOCA_H
 # include <alloca.h>
+# endif
 # include <string.h>
 # define strndupa(s, n)                                                       \
     ({                                                                        \
--- handlers.c.orig
+++ handlers.c
@@ -38,7 +38,9 @@
 #include <signal.h>
 #include <assert.h>
 #include <setjmp.h>
+#ifdef HAVE_ALLOCA_H
 #include <alloca.h>
+#endif
 #include <strings.h>
 #include <string.h>
 #include <stdbool.h>
--- maps.c.orig
+++ maps.c
@@ -31,7 +31,9 @@
 #include <stddef.h>
 #include <stdlib.h>
 #include <string.h>
+#ifdef HAVE_ALLOCA_H
 #include <alloca.h>
+#endif
 #include <stdbool.h>
 #include <unistd.h>
 
```
This works on my FreeBSD system, and I don't see why it should not work elsewhere.
